src/parquet/file/file-metadata-test.cc

```diff
@@ -0,0 +1,178 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gtest/gtest.h>
+#include "parquet/file/metadata.h"
+#include "parquet/schema/types.h"
+#include "parquet/types.h"
+
+namespace parquet {
+
+namespace metadata {
+
+TEST(MetadataReaderWriter, TestReadWrite) {
+  std::shared_ptr<WriterProperties> props = WriterProperties::Builder().build();
+
+  int64_t nrows = 1000;
+  std::unique_ptr<FileMetaDataWriter> f_writer =
+      FileMetaDataWriter::GetFileMetaDataWriter(*props.get());
+  std::unique_ptr<SchemaMetaDataWriter> s_writer1 =
+      SchemaMetaDataWriter::GetSchemaMetaDataWriter();
+  std::unique_ptr<SchemaMetaDataWriter> s_writer2 =
+      SchemaMetaDataWriter::GetSchemaMetaDataWriter();
+  std::unique_ptr<SchemaMetaDataWriter> s_writer3 =
+      SchemaMetaDataWriter::GetSchemaMetaDataWriter();
+  std::unique_ptr<RowGroupMetaDataWriter> rg_writer =
+      RowGroupMetaDataWriter::GetRowGroupMetaDataWriter(*props.get());
+  std::shared_ptr<schema::ColumnPath> col_path0 =
+      schema::ColumnPath::FromDotString("float_col");
+  std::shared_ptr<schema::ColumnPath> col_path1 =
+      schema::ColumnPath::FromDotString("int_col");
+
+  ColumnStatistics stats_int;
+  stats_int.null_count = 0;
+  stats_int.distinct_count = nrows;
+  std::string int_min = std::string("100");
+  std::string int_max = std::string("200");
+  stats_int.min = &int_min;
+  stats_int.max = &int_max;
+  ColumnStatistics stats_float;
+  stats_float.null_count = 0;
+  stats_float.distinct_count = nrows;
+  std::string float_min = std::string("100.100");
+  std::string float_max = std::string("200.200");
+  stats_float.min = &float_min;
+  stats_float.max = &float_max;
+
+  // Write the metadata
+  // file metadata
+  f_writer->set_num_rows(nrows);
+  // rowgroup metadata
+  rg_writer->initialize_columns(2);
+  rg_writer->set_num_rows(nrows);
+  rg_writer->set_total_byte_size(1024);
+  // column metadata
+  rg_writer->set_column_file_offset(0, 0);
+  rg_writer->set_column_file_offset(1, 1024);
+  rg_writer->set_column_type(0, Type::FLOAT);
+  rg_writer->set_column_type(1, Type::INT32);
+  rg_writer->set_column_encodings(0, col_path0);
+  rg_writer->set_column_encodings(1, col_path1);
+  rg_writer->set_column_path_in_schema(0, col_path0);
+  rg_writer->set_column_path_in_schema(1, col_path1);
+  rg_writer->set_column_codecs(0, col_path0);
+  rg_writer->set_column_codecs(1, col_path1);
+  rg_writer->set_column_num_values(0, nrows);
+  rg_writer->set_column_num_values(1, nrows);
+  rg_writer->set_column_total_uncompressed_size(0, 512);
+  rg_writer->set_column_total_uncompressed_size(1, 512);
+  rg_writer->set_column_total_compressed_size(0, 512);
+  rg_writer->set_column_total_compressed_size(1, 512);
```

Comment:
I think we should split this out into the required and optional `ColumnMetaData` fields

https://github.com/apache/parquet-format/blob/master/src/main/thrift/parquet.thrift#L466

For example

```c++
auto col_meta_builder = rg_meta_builder->AddColumn(type, encodings, path_in_schema, 
    codec, num_values, uncompressed_size, compressed_size, data_page_offset)
```

Then add methods that set the optional fields individually (for example: key-value metadata). This will also reduce errors from user failure to provide a value for a required field